### PR TITLE
8287132: Retire Runtime.runFinalizersOnExit so that it always throws UOE

### PR DESCRIPTION
**Pull request by dholmes-ora**
Please review these further changes for JDK 8 Maintenance Release 4 - JSR-337.

The MR4 changes to the reference processing implementation have made continued support for `runFinalizersonExit` unreasonable and so it is "retired" and will always throw `UnsupportedOperationException`. (This is what was also done for `Thread.stop(Throwable)` in JDK 8.)

Various updates to the specifications are made to remove references to running finalizers at exit - see the CSR request for details.

All dead code (including the already dead `JVM_Exit`) is removed. Shutdown.java is replaced by the version from mainline with a slight modification as 8u doesn't have `sun.misc.VM.isShutdown()`.

Tests using RFOE are deleted and a new test to check the throwing of UOE added. The new test was NOT a rename of an existing test but for some reason git is treating it that way.

Testing: :jdk_core group

Thanks,
David

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8287132](https://bugs.openjdk.java.net/browse/JDK-8287132): Retire Runtime.runFinalizersOnExit so that it always throws UOE
 * [JDK-8287133](https://bugs.openjdk.java.net/browse/JDK-8287133): Retire Runtime.runFinalizersOnExit so that it always throws UOE (**CSR**)

### Reviewers
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [3de362a0](https://git.openjdk.java.net/jdk8u-ri/pull/9/files/3de362a0c950e2a7037d5998c52370e20ae4ae29)
 * [Iris Clark](https://openjdk.java.net/census#iris) (@irisclark - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk8u-ri pull/9/head:pull/9` \
`$ git checkout pull/9`

Update a local copy of the PR: \
`$ git checkout pull/9` \
`$ git pull https://git.openjdk.java.net/jdk8u-ri pull/9/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9`

View PR using the GUI difftool: \
`$ git pr show -t 9`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk8u-ri/pull/9.diff">https://git.openjdk.java.net/jdk8u-ri/pull/9.diff</a>

</details>

### CODE REVIEW

File: hotspot/src/share/vm/runtime/thread.cpp [hunk 1]
```diff
@@ -3916,10 +3916,9 @@ void JavaThread::invoke_shutdown_hooks() {
     // SystemDictionary::resolve_or_null will return null if there was
     // an exception.  If we cannot load the Shutdown class, just don't
     // call Shutdown.shutdown() at all.  This will mean the shutdown hooks
-    // and finalizers (if runFinalizersOnExit is set) won't be run.
-    // Note that if a shutdown hook was registered or runFinalizersOnExit
-    // was called, the Shutdown class would have already been loaded
-    // (Runtime.addShutdownHook and runFinalizersOnExit will load it).
+    // won't be run. Note that if a shutdown hook was registered,
+    // the Shutdown class would have already been loaded
+    // (Runtime.addShutdownHook will load it).
```
Comment (mlchung) [inline hotspot/src/share/vm/runtime/thread.cpp:3921]: Nit: The comment block about "Shutdown sequence" below has a reference to finalization-on-exit.

```
3945           // shutdown hooks, run finalizers if finalization-on-exit
```

File: jdk/src/share/classes/java/lang/Shutdown.java [hunk 2]
```diff
@@ -121,13 +128,20 @@ private static void runHooks() {
...
                     ThreadDeath td = (ThreadDeath)t;
                     throw td;
                 }
             }
         }
+
+        // set shutdown state
+        // Synchronization is for visibility; only one thread
+        // can ever get here.
+        synchronized(lock) {
```
Comment (mlchung) [inline jdk/src/share/classes/java/lang/Shutdown.java]:
```suggestion
        synchronized (lock) {
```

Nit: a space after `synchronized`, consistent with the style in this local file.

**Review by mlchung — APPROVED** [review]
Looks good.   I verified this PR matches the implementation by JDK-8198249 dropping finalization-on-exit.

File: hotspot/src/share/vm/runtime/thread.cpp (hunk 1 above)
Comment (dholmes-ora) [reply hotspot/src/share/vm/runtime/thread.cpp:3921]: Thanks, that escaped my grep search. I will fix.

File: jdk/src/share/classes/java/lang/Shutdown.java (hunk 2 above)
Comment (dholmes-ora) [reply jdk/src/share/classes/java/lang/Shutdown.java]: Fixed.

**Review by irisclark — APPROVED** [review]
Spec changes in java.lang.{Runtime,Shutdown,System} align with CSR 8287133 and are as expected.

**Comment by mlchung** [comment]
As described in CSR 8287133, `Runtime.runFinalizersOnExit` was removed from Java 11 in 2018.   We did the corpus analysis at that time showed very few uses, and most of those were gone in later versions of the same software. A rerun of the corpus analysis done in 2022 found only a single direct use of `runFinalizersOnExit` and that was only in non-product builds.  It is believed that retiring `Runtime.runFinalizersOnExit` will have minimal real world impact.

**Pull request closed without merging**